cmd/dep/gb_importer.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func (i *gbImporter) Name() string {
+	return "gb"
+}
+
+func (i *gbImporter) HasDepMetadata(dir string) bool {
+	// gb stores the manifest in the vendor tree
+	var m = filepath.Join(dir, "vendor", "manifest")
+	if _, err := os.Stat(m); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (i *gbImporter) Import(dir string, pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	err := i.load(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return i.convert(pr)
+}
+
+// load the gb manifest
+func (i *gbImporter) load(projectDir string) error {
+	i.logger.Println("Detected gb manifest file...")
+	var mf = filepath.Join(projectDir, "vendor", "manifest")
+	if i.verbose {
+		i.logger.Printf("  Loading %s", mf)
+	}
+
+	var buf []byte
+	var err error
+	if buf, err = ioutil.ReadFile(mf); err != nil {
+		return errors.Wrapf(err, "Unable to read %s", mf)
+	}
+	if err := json.Unmarshal(buf, &i.manifest); err != nil {
+		return errors.Wrapf(err, "Unable to parse %s", mf)
+	}
+
+	return nil
+}
+
+// convert the gb manifest into dep configuration files.
+func (i *gbImporter) convert(pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	i.logger.Println("Converting from gb manifest...")
+
+	manifest := &dep.Manifest{
+		Constraints: make(gps.ProjectConstraints),
+	}
+
+	lock := &dep.Lock{}
+
+	for _, pkg := range i.manifest.Dependencies {
+		pc, lp, err := i.convertOne(pkg)
+		if err != nil {
+			return nil, nil, err
+		}
+		manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{Source: pc.Ident.Source, Constraint: pc.Constraint}
+		lock.P = append(lock.P, lp)
+	}
+
+	return manifest, lock, nil
+}
+
+func (i *gbImporter) convertOne(pkg gbDependency) (pc gps.ProjectConstraint, lp gps.LockedProject, err error) {
+	if pkg.Importpath == "" {
+		err = errors.New("Invalid gb configuration, package import path is required")
+		return
+	}
+
+	/*
+		gb's vendor plugin (gb vendor), which manages the vendor tree and manifest
+		file, supports fetching by a specific tag or revision, but if you specify
+		either of those it's a detached checkout and the "branch" field is HEAD.
+		The only time the "branch" field is not "HEAD" is if you do not specify a
+		tag or revision, otherwise it's either "master" or the value of the -branch
+		flag
+
+		This means that, generally, the only possible "constraint" we can really specify is
+		the branch name if the branch name is not HEAD. Otherwise, it's a specific revision.
+
+		However, if we can infer a tag that points to the revision or the branch, we may be able
+		to use that as the constraint
+	*/
+
+	pc.Ident = gps.ProjectIdentifier{ProjectRoot: gps.ProjectRoot(pkg.Importpath), Source: pkg.Repository}
+
+	// Generally, gb tracks revisions
+	var revision = gps.Revision(pkg.Revision)
+
+	// But if the branch field is not "HEAD", we can use that as the initial constraint
+	var constraint gps.Constraint
+	if pkg.Branch != "HEAD" {
+		constraint = gps.NewBranch(pkg.Branch)
```

Comment:
This method is missing a conditional branch that uses the constraint if a branch is set, and otherwise uses the result of `sm.InferConstraint`. Right now, constraint is set to master, then immediately discarded a few lines down.

Here are the steps we need:
1. Default the constraint to the branch if possible.
1. Use that constraint to lookup the version to put in the lock.
1. If the constraint is empty, try to infer it from the locked version.